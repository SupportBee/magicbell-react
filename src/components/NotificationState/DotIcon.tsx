import React from 'react';

/**
 * SVG circle.
 *
 * @example
 * <DotIcon  />
 */
export default function DotIcon() {
  return (
    <svg viewBox="0 0 512 512">
      <circle cx="256" cy="256" r="64" />
    </svg>
  );
}
