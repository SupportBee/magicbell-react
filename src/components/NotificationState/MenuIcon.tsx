import React from 'react';

/**
 * SVG with 3 vertical dots.
 *
 * @example
 * <MenuIcon  />
 */
export default function MenuIcon() {
  return (
    <svg viewBox="0 0 512 512">
      <circle cx="256" cy="64" r="64" />
      <circle cx="256" cy="256" r="64" />
      <circle cx="256" cy="448" r="64" />
    </svg>
  );
}
