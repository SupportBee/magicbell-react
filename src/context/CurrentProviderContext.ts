import { createContext } from 'react';

export const NO_PROVIDER = 'MAGICBELL_NO_PROVIDER';

const CurrentProviderContext = createContext<String>(NO_PROVIDER);
export default CurrentProviderContext;
