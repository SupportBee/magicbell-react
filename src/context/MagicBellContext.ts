import { createContext, useContext } from 'react';

export interface IMagicBellContext {
  images?: Partial<{
    emptyInboxUrl: string;
  }>;
}

const MagicBellContext = createContext<IMagicBellContext>({} as IMagicBellContext);
export default MagicBellContext;

const useMagicBellContext = () => useContext(MagicBellContext);
export { useMagicBellContext };
