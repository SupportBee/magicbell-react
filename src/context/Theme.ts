export interface IElementTheme {
  backgroundColor: string;
  backgroundOpacity: number;
  borderRadius: string;
  fontFamily: string;
  fontSize: string;
  textAlign: 'left' | 'right' | 'center' | 'justify' | 'initial' | 'inherit';
  textColor: string;
  textTransform: 'none' | 'capitalize' | 'uppercase' | 'lowercase' | 'initial' | 'inherit';
}

export interface IMagicBellTheme {
  icon: {
    borderColor: string;
    width: string;
  };
  unseenBadge: IElementTheme;
  header: IElementTheme;
  container: IElementTheme;
  footer: IElementTheme;
  notification: {
    default: IElementTheme;
    unread: IElementTheme;
    unseen: IElementTheme;
  };
}

const blue = '#3498F4';
const red = '#DF4759';
const black = '#3A424D';

// prettier-ignore
const fontFamily = '-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen-Sans, Ubuntu, Cantarell, "Helvetica Neue", sans-serif';
const fontSize = '14px';
const textAlign = 'left';
const textTransform = 'none';

export const defaultTheme = {
  icon: {
    borderColor: blue,
    width: '24px',
  },
  header: {
    backgroundColor: blue,
    backgroundOpacity: 1,
    borderRadius: '8px',
    fontFamily,
    fontSize,
    textAlign,
    textColor: 'white',
    textTransform,
  },
  footer: {
    backgroundColor: blue,
    backgroundOpacity: 1,
    borderRadius: '8px',
    fontFamily,
    fontSize: fontSize,
    textAlign,
    textColor: 'white',
    textTransform,
  },
  unseenBadge: {
    backgroundColor: red,
    backgroundOpacity: 1,
    borderRadius: '2px',
    fontFamily,
    fontSize: '10px',
    textAlign: 'center',
    textColor: 'white',
    textTransform,
  },
  container: {
    backgroundColor: '#FFFFFF',
    backgroundOpacity: 1,
    borderRadius: '8px',
    fontFamily,
    fontSize: '14px',
    textAlign,
    textColor: black,
    textTransform,
  },
  notification: {
    default: {
      backgroundColor: blue,
      backgroundOpacity: 0.1,
      borderRadius: '8px',
      fontFamily,
      fontSize,
      textAlign,
      textColor: black,
      textTransform,
    },
    unread: {
      backgroundColor: blue,
      backgroundOpacity: 0,
      borderRadius: '8px',
      fontFamily,
      fontSize,
      textAlign,
      textColor: black,
      textTransform,
    },
    unseen: {
      backgroundColor: blue,
      backgroundOpacity: 0.05,
      borderRadius: '8px',
      fontFamily,
      fontSize,
      textAlign,
      textColor: black,
      textTransform,
    },
  },
} as IMagicBellTheme;
