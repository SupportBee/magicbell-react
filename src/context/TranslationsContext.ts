import { createContext, useContext } from 'react';

const TranslationsContext = createContext<any>({});
export default TranslationsContext;

const TranslationsProvider = TranslationsContext.Provider;
const useTranslations = () => useContext(TranslationsContext);
export { TranslationsProvider, useTranslations };
