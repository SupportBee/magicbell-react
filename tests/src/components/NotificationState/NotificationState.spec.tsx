import { useNotificationFactory } from '@magicbell/react-headless';
import { render, RenderResult, waitFor } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';
import React from 'react';
import NotificationState from '../../../../src/components/NotificationState';
import { MagicBellThemeProvider } from '../../../../src/context/MagicBellThemeContext';
import { defaultTheme } from '../../../../src/context/Theme';
import { sampleNotification } from '../../../factories/NotificationFactory';

describe('components', () => {
  describe('NotificationState', () => {
    let view: RenderResult;

    beforeEach(() => {
      const { result } = renderHook(() => useNotificationFactory(sampleNotification));

      view = render(
        <MagicBellThemeProvider value={defaultTheme}>
          <NotificationState notification={result.current} />
        </MagicBellThemeProvider>,
      );
    });

    describe('render', () => {
      describe('the notification is read', () => {
        beforeEach(() => {
          const { result } = renderHook(() =>
            useNotificationFactory({ ...sampleNotification, readAt: Date.now() }),
          );

          view.rerender(
            <MagicBellThemeProvider value={defaultTheme}>
              <NotificationState notification={result.current} />
            </MagicBellThemeProvider>,
          );
        });

        it('renders an icon for the context menu', () => {
          expect(view.container).toMatchSnapshot();
        });
      });

      describe('the notification is unread', () => {
        beforeEach(() => {
          const { result } = renderHook(() =>
            useNotificationFactory({ ...sampleNotification, readAt: null }),
          );

          view.rerender(
            <MagicBellThemeProvider value={defaultTheme}>
              <NotificationState notification={result.current} />
            </MagicBellThemeProvider>,
          );
        });

        it('renders a dot', () => {
          expect(view.container).toMatchSnapshot();
        });
      });
    });

    describe('.handleClick', () => {
      it('renders the context menu', async () => {
        const bell = view.getByTestId('icon');
        userEvent.click(bell);

        await waitFor(() => expect(view.container).toMatchSnapshot());
      });

      describe('the menuPlacement is set to "top-end"', () => {
        beforeEach(() => {
          const { result } = renderHook(() => useNotificationFactory(sampleNotification));

          view.rerender(
            <MagicBellThemeProvider value={defaultTheme}>
              <NotificationState notification={result.current} menuPlacement="top-end" />
            </MagicBellThemeProvider>,
          );
        });

        it('renders a menu in the specified position', async () => {
          userEvent.click(view.getByTestId('icon'));

          expect(await view.findByText('Delete')).toBeInTheDocument();
          await waitFor(() => expect(view.container).toMatchSnapshot());
        });
      });
    });

    describe('.onHover', () => {
      describe('the notification is unread', () => {
        it('renders the menu icon', async () => {
          userEvent.hover(view.getByTestId('icon'));

          await waitFor(() => expect(view.container).toMatchSnapshot());
        });
      });

      describe('.onBlur', () => {
        it('renders the dot icon', async () => {
          userEvent.hover(view.getByTestId('icon'));
          userEvent.unhover(view.getByTestId('icon'));

          await waitFor(() => expect(view.container).toMatchSnapshot());
        });
      });
    });
  });
});
