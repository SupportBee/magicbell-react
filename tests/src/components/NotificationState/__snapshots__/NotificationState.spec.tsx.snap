// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components NotificationState .handleClick renders the context menu 1`] = `
.emotion-0 {
  display: -webkit-box!important;
  display: -webkit-flex!important;
  display: -ms-flexbox!important;
  display: flex!important;
  -webkit-align-items: center!important;
  -webkit-box-align: center!important;
  -ms-flex-align: center!important;
  align-items: center!important;
  padding: 0.5em 0 0.5em 0.25em!important;
  color: #3498F4!important;
}

.emotion-0 svg {
  height: 1.5em;
  fill: currentColor;
}

.emotion-1 {
  background: #FFFFFF!important;
  border-radius: 8px!important;
  box-shadow: 2px 2px 3px rgba(230, 230, 230, 0.55);
  color: #3A424D!important;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Oxygen-Sans,Ubuntu,Cantarell,"Helvetica Neue",sans-serif!important;
  font-size: 14px!important;
  text-transform: none!important;
  min-width: 10em;
}

.emotion-1 a {
  display: block;
  padding: 0.75em 1.25em!important;
}

.emotion-1 a:hover {
  -webkit-text-decoration: underline!important;
  text-decoration: underline!important;
}

<div>
  <div
    aria-expanded="false"
    class="emotion-0"
    data-testid="icon"
  >
    <svg
      viewBox="0 0 512 512"
    >
      <circle
        cx="256"
        cy="64"
        r="64"
      />
      <circle
        cx="256"
        cy="256"
        r="64"
      />
      <circle
        cx="256"
        cy="448"
        r="64"
      />
    </svg>
  </div>
  <div
    data-tippy-root=""
    id="tippy-3"
    style="z-index: 0; transition: none; position: absolute; left: 0px; top: 0px; margin: 0px; pointer-events: none;"
  >
    <div
      class="emotion-1"
    >
      <a
        data-testid="toggle-read"
      >
        Mark as read
      </a>
      <a
        data-testid="delete"
      >
        Delete
      </a>
    </div>
  </div>
</div>
`;

exports[`components NotificationState .handleClick the menuPlacement is set to "top-end" renders a menu in the specified position 1`] = `
.emotion-0 {
  display: -webkit-box!important;
  display: -webkit-flex!important;
  display: -ms-flexbox!important;
  display: flex!important;
  -webkit-align-items: center!important;
  -webkit-box-align: center!important;
  -ms-flex-align: center!important;
  align-items: center!important;
  padding: 0.5em 0 0.5em 0.25em!important;
  color: #3498F4!important;
}

.emotion-0 svg {
  height: 1.5em;
  fill: currentColor;
}

.emotion-1 {
  background: #FFFFFF!important;
  border-radius: 8px!important;
  box-shadow: 2px 2px 3px rgba(230, 230, 230, 0.55);
  color: #3A424D!important;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Oxygen-Sans,Ubuntu,Cantarell,"Helvetica Neue",sans-serif!important;
  font-size: 14px!important;
  text-transform: none!important;
  min-width: 10em;
}

.emotion-1 a {
  display: block;
  padding: 0.75em 1.25em!important;
}

.emotion-1 a:hover {
  -webkit-text-decoration: underline!important;
  text-decoration: underline!important;
}

<div>
  <div
    aria-expanded="false"
    class="emotion-0"
    data-testid="icon"
  >
    <svg
      viewBox="0 0 512 512"
    >
      <circle
        cx="256"
        cy="64"
        r="64"
      />
      <circle
        cx="256"
        cy="256"
        r="64"
      />
      <circle
        cx="256"
        cy="448"
        r="64"
      />
    </svg>
  </div>
  <div
    data-tippy-root=""
    id="tippy-4"
    style="z-index: 0; transition: none; pointer-events: none; position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; bottom: 0px; transform: translate(0px, -2px);"
  >
    <div
      class="emotion-1"
    >
      <a
        data-testid="toggle-read"
      >
        Mark as read
      </a>
      <a
        data-testid="delete"
      >
        Delete
      </a>
    </div>
  </div>
</div>
`;

exports[`components NotificationState .onHover .onBlur renders the dot icon 1`] = `
.emotion-0 {
  display: -webkit-box!important;
  display: -webkit-flex!important;
  display: -ms-flexbox!important;
  display: flex!important;
  -webkit-align-items: center!important;
  -webkit-box-align: center!important;
  -ms-flex-align: center!important;
  align-items: center!important;
  padding: 0.5em 0 0.5em 0.25em!important;
  color: #3498F4!important;
}

.emotion-0 svg {
  height: 1.5em;
  fill: currentColor;
}

.emotion-1 {
  background: #FFFFFF!important;
  border-radius: 8px!important;
  box-shadow: 2px 2px 3px rgba(230, 230, 230, 0.55);
  color: #3A424D!important;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Oxygen-Sans,Ubuntu,Cantarell,"Helvetica Neue",sans-serif!important;
  font-size: 14px!important;
  text-transform: none!important;
  min-width: 10em;
}

.emotion-1 a {
  display: block;
  padding: 0.75em 1.25em!important;
}

.emotion-1 a:hover {
  -webkit-text-decoration: underline!important;
  text-decoration: underline!important;
}

<div>
  <div
    aria-expanded="true"
    class="emotion-0"
    data-testid="icon"
  >
    <svg
      viewBox="0 0 512 512"
    >
      <circle
        cx="256"
        cy="256"
        r="64"
      />
    </svg>
  </div>
  <div
    data-tippy-root=""
    id="tippy-6"
    style="z-index: 0; transition: none; position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; transform: translate(0px, 2px);"
  >
    <div
      class="emotion-1"
    >
      <a
        data-testid="toggle-read"
      >
        Mark as read
      </a>
      <a
        data-testid="delete"
      >
        Delete
      </a>
    </div>
  </div>
</div>
`;

exports[`components NotificationState .onHover the notification is unread renders the menu icon 1`] = `
.emotion-0 {
  display: -webkit-box!important;
  display: -webkit-flex!important;
  display: -ms-flexbox!important;
  display: flex!important;
  -webkit-align-items: center!important;
  -webkit-box-align: center!important;
  -ms-flex-align: center!important;
  align-items: center!important;
  padding: 0.5em 0 0.5em 0.25em!important;
  color: #3498F4!important;
}

.emotion-0 svg {
  height: 1.5em;
  fill: currentColor;
}

.emotion-1 {
  background: #FFFFFF!important;
  border-radius: 8px!important;
  box-shadow: 2px 2px 3px rgba(230, 230, 230, 0.55);
  color: #3A424D!important;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Oxygen-Sans,Ubuntu,Cantarell,"Helvetica Neue",sans-serif!important;
  font-size: 14px!important;
  text-transform: none!important;
  min-width: 10em;
}

.emotion-1 a {
  display: block;
  padding: 0.75em 1.25em!important;
}

.emotion-1 a:hover {
  -webkit-text-decoration: underline!important;
  text-decoration: underline!important;
}

<div>
  <div
    aria-expanded="false"
    class="emotion-0"
    data-testid="icon"
  >
    <svg
      viewBox="0 0 512 512"
    >
      <circle
        cx="256"
        cy="64"
        r="64"
      />
      <circle
        cx="256"
        cy="256"
        r="64"
      />
      <circle
        cx="256"
        cy="448"
        r="64"
      />
    </svg>
  </div>
  <div
    data-tippy-root=""
    id="tippy-5"
    style="z-index: 0; transition: none; position: absolute; left: 0px; top: 0px; margin: 0px;"
  >
    <div
      class="emotion-1"
    >
      <a
        data-testid="toggle-read"
      >
        Mark as read
      </a>
      <a
        data-testid="delete"
      >
        Delete
      </a>
    </div>
  </div>
</div>
`;

exports[`components NotificationState render the notification is read renders an icon for the context menu 1`] = `
.emotion-0 {
  display: -webkit-box!important;
  display: -webkit-flex!important;
  display: -ms-flexbox!important;
  display: flex!important;
  -webkit-align-items: center!important;
  -webkit-box-align: center!important;
  -ms-flex-align: center!important;
  align-items: center!important;
  padding: 0.5em 0 0.5em 0.25em!important;
  color: rgba(58, 66, 77, 0.5)!important;
}

.emotion-0 svg {
  height: 1.5em;
  fill: currentColor;
}

<div>
  <div
    aria-expanded="false"
    class="emotion-0"
    data-testid="icon"
  >
    <svg
      viewBox="0 0 512 512"
    >
      <circle
        cx="256"
        cy="64"
        r="64"
      />
      <circle
        cx="256"
        cy="256"
        r="64"
      />
      <circle
        cx="256"
        cy="448"
        r="64"
      />
    </svg>
  </div>
</div>
`;

exports[`components NotificationState render the notification is unread renders a dot 1`] = `
.emotion-0 {
  display: -webkit-box!important;
  display: -webkit-flex!important;
  display: -ms-flexbox!important;
  display: flex!important;
  -webkit-align-items: center!important;
  -webkit-box-align: center!important;
  -ms-flex-align: center!important;
  align-items: center!important;
  padding: 0.5em 0 0.5em 0.25em!important;
  color: #3498F4!important;
}

.emotion-0 svg {
  height: 1.5em;
  fill: currentColor;
}

<div>
  <div
    aria-expanded="false"
    class="emotion-0"
    data-testid="icon"
  >
    <svg
      viewBox="0 0 512 512"
    >
      <circle
        cx="256"
        cy="256"
        r="64"
      />
    </svg>
  </div>
</div>
`;
